### PR TITLE
Patch langchain async APIs for autologging

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12114?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12114/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12114
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Patch `ainvoke, abatch, astream` methods for runnables.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

Had a test on the patching:
https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/526657760899964
TLDR since autologging is not async, patching async functions will make the inference call non-async as well. For example if you have 10 inference jobs, and injecting mlflow callback takes 1 second, then total time increase would be 10 seconds instead of 1 second. 
Mlflowcallback injection won't take that long though, but logging artifacts would become a performance concern. I added a note in the docstring, so the suggestion is that if users really want to enable tracing using autologging, then they should disable logging any artifacts -- it's off by default.

**Update**: test in https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/2393426062477360/command/1344703767318006 
Conclusion is that autologging patching async methods introduces a ~0.5 second overhead on each task, but the tasks themselves are still async, so the total overhead = 0.5 * number_of_async_tasks. 
User can opt-out this behavior by setting `mlflow.langchaing.autolog(disable=True)`. 
By default mlflow langchain autologging patches all invoke methods for sync and async cases.

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/utils/openai_utils.py [hunk 1]
```diff
@@ -166,6 +167,26 @@ def _mock_request(**kwargs):
         yield m
 
 
+@contextmanager
+def _mock_openai_arequest(**kwargs):
+    with mock.patch("aiohttp.ClientSession.request", **kwargs) as mock_request:
+        mock_request.side_effect = _mock_async_chat_completion_response
+        yield mock_request
```
Comment (harupy) [inline mlflow/utils/openai_utils.py:174]:
```suggestion
    with mock.patch("aiohttp.ClientSession.request", **kwargs, return_value=_mock_async_chat_completion_response) as mock_request:
        yield mock_request
```

File: mlflow/utils/openai_utils.py [hunk 2]
```diff
@@ -166,6 +167,26 @@ def _mock_request(**kwargs):
...
+
+
+async def _mock_async_chat_completion_response(content=TEST_CONTENT, **kwargs):
```
Comment (harupy) [inline mlflow/utils/openai_utils.py:177]: Is `kwargs` used?

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -929,7 +929,7 @@ def _patch_runnable_cls(cls):
     For classes that are subclasses of Runnable, we patch the `invoke`, `batch`, and `stream`
     methods for autologging.
     """
-    for func_name in ["invoke", "batch", "stream"]:
+    for func_name in ["invoke", "batch", "stream", "ainvoke", "abatch", "astream"]:
```
Comment (harupy) [inline mlflow/langchain/__init__.py:932]: I'm not sure if it's really safe to inject autologging logic into an **_async_** function.

Comment (harupy) [reply mlflow/langchain/__init__.py:932]: Just because the result seems correct doesn't mean it's safe. If autologging contains blocking operations, I think that would affect the performance.

Comment (serena-ruan) [reply mlflow/langchain/__init__.py:932]: What's the concern?

Comment (harupy) [reply mlflow/langchain/__init__.py:932]: Can we create an example like below and check how it behaves and how long it takes with/without autlogging enabled?

```
t1 = runnable.ainvoke(...)
t2 = runnable.ainvoke(...)
t3 = runnable.ainvoke(...)
asyncio.gather(t1, t2, t3)
```

Comment (harupy) [reply mlflow/langchain/__init__.py:932]: The concern is blocking operations such as an HTTP request by requests or urllib3.

File: mlflow/utils/openai_utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/openai_utils.py:174]: Wait, is `_mock_async_chat_completion_response` a function?

File: mlflow/langchain/__init__.py (hunk 3 above)
Comment (harupy) [reply mlflow/langchain/__init__.py:932]:
```python
import asyncio
import time

async def test1():
    for _ in range(0, 3):
        print('Test1')
        time.sleep(1) # Here
        
async def test2():
    for _ in range(0, 3):
        print('Test2')
        time.sleep(1) # Here
    
async def main():
    await asyncio.gather(test1(), test2()) # Here

asyncio.run(main())
```

is a good example. `time.sleep` (blocking operation) makes the function non async. I think the same would happen if we replace it with `requests.get`.

**Comment by harupy** [comment]
There is another concern. In autologging, a patched looks like this:

```
def patched(*args, **kwargs):
   pre_call(...)
   res = original_func(...)
   post_call(...)
```

If `original_func` is an async func, there is no guarantee that `original_func` has finished when `post_call` is called.

File: mlflow/langchain/__init__.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/langchain/__init__.py]: image

**Comment by serena-ruan** [comment]
> There is another concern. In autologging, a patched function looks like this:
> 
> ```python
> def patched(*args, **kwargs):
>    pre_call(...)
>    res = original_func(...)
>    post_call(...)
> ```
> 
> If `original_func` is an async func, there is no guarantee that `original_func` has finished when `post_call` is called because there is no `await`.

It seems fine as long as post_call doesn't depend on the result right? BTW I added `allow_async` param in `autolog` to control whether we patch async methods or not

**Comment by harupy** [comment]
> It seems fine as long as post_call doesn't depend on the result right? BTW I added allow_async param in autolog to control whether we patch async methods or not

Correct

File: mlflow/langchain/__init__.py (hunk 3 above)
Comment (harupy) [reply mlflow/langchain/__init__.py]: Can we insert the following in `patched_inference` and check what happens?

```
time.sleep(1)
print("foo")
```

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -1025,6 +1039,8 @@ def autolog(
             We do not guarantee classes specified in this list can be logged as a model, but tracing
             will be supported. Note that all classes within the list must be subclasses of Runnable,
             and we only patch `invoke`, `batch`, and `stream` methods for tracing.
+        allow_async: If ``True``, enable autologging on langchain model async methods. Default to
+            ``False``.
```
Comment (harupy) [inline mlflow/langchain/__init__.py:1043]: Should we describe the risk of turning this on?

File: mlflow/langchain/__init__.py
Comment (harupy) [inline mlflow/langchain/__init__.py:1]: /ai suggest names for the `allow_async` argument

File: mlflow/langchain/__init__.py [hunk 5]
```diff
@@ -1056,6 +1060,12 @@ def autolog(
             We do not guarantee classes specified in this list can be logged as a model, but tracing
             will be supported. Note that all classes within the list must be subclasses of Runnable,
             and we only patch `invoke`, `batch`, and `stream` methods for tracing.
+        allow_async: If ``True``, enable autologging on langchain model async methods. Default to
+            ``False``. Note that if this is enabled and any of the following options are turned on
+            (log_models, log_input_examples, log_model_signatures, or log_inputs_outputs),
+            autologging may cause blocking in the model inference process because autologging itself
+            is not asynchronous. If no artifact logging is enabled, autologging should have minimal
+            or no impact on the async execution of model inference.
```
Comment (harupy) [inline mlflow/langchain/__init__.py:1068]:
```suggestion
            on the async execution of model inference.
```

I'd remove `no impatc` because there must be some impact.

File: mlflow/langchain/__init__.py [hunk 6]
```diff
@@ -1056,6 +1060,12 @@ def autolog(
             We do not guarantee classes specified in this list can be logged as a model, but tracing
             will be supported. Note that all classes within the list must be subclasses of Runnable,
             and we only patch `invoke`, `batch`, and `stream` methods for tracing.
+        allow_async: If ``True``, enable autologging on langchain model async methods. Default to
+            ``False``. Note that if this is enabled and any of the following options are turned on
+            (log_models, log_input_examples, log_model_signatures, or log_inputs_outputs),
+            autologging may cause blocking in the model inference process because autologging itself
```
Comment (harupy) [inline mlflow/langchain/__init__.py:1066]:
```suggestion
            autologging may slow down the model inference process because autologging itself
```

can we replace `blocking` with slow down?

File: mlflow/langchain/__init__.py
Comment (harupy) [reply mlflow/langchain/__init__.py:1]: @serena-ruan What about `patch_async_methods`?

Comment (serena-ruan) [reply mlflow/langchain/__init__.py:1]: I think this makes sense to devs, but I'm not sure if it makes sense to customers 🤔 Do they really know what we're doing with `patch`

Comment (serena-ruan) [reply mlflow/langchain/__init__.py:1]: Let's vote in our slack channel 😆

**Comment by harupy** [comment]
@serena-ruan Is it possible to create an async langchain example to assess the impact more accurately?

File: tests/langchain/test_langchain_autolog.py [hunk 7]
```diff
@@ -785,6 +789,37 @@ def test_langchain_autolog_callback_injection_in_invoke(invoke_arg, generate_cal
         assert callbacks[0].logs == ["chain_start", "chain_end"]
 
 
+@pytest.mark.parametrize("invoke_arg", ["args", "kwargs"])
+@pytest.mark.parametrize(
+    "generate_callbacks",
+    [lambda: [CustomCallbackHandler()], lambda: BaseCallbackManager([CustomCallbackHandler()])],
```
Comment (harupy) [inline tests/langchain/test_langchain_autolog.py:795]: can we also test `AsyncCallbackHandler`?

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/12114/files#r1637856802 is addressed!

File: tests/langchain/test_langchain_autolog.py (hunk 7 above)
Comment (serena-ruan) [reply tests/langchain/test_langchain_autolog.py]: added test!
